### PR TITLE
fix catalog performance issue

### PR DESCRIPTION
**Pull request by wy65701436**
Signed-off-by: Wang Yan <wangyan@vmware.com>

### CODE REVIEW

File: src/server/registry/catalog.go [hunk 1]
```diff
@@ -68,16 +75,50 @@ func (r *repositoryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
...
+		r.wp.GetWorker()
+		go func(repo *models.RepoRecord) {
+			defer func() {
+				r.wp.ReleaseWorker()
+				wg.Done()
+			}()
+
+			valid, err := r.validateRepo(orm.Clone(req.Context()), repo.RepositoryID)
+			if err != nil {
+				lib_http.SendError(w, err)
```
Comment (reasonerjt) [inline src/server/registry/catalog.go:109]: Use a channel to cancel the loop?

**Comment by reasonerjt** [comment]
Once this is done, may consider to CP to 2.1.x?

File: src/server/registry/catalog.go [hunk 2]
```diff
@@ -68,16 +75,50 @@ func (r *repositoryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
...
+
+	for _, repo := range repoRecords {
+		r.wp.GetWorker()
+		go func(repo *models.RepoRecord) {
+			defer func() {
+				r.wp.ReleaseWorker()
+				wg.Done()
+			}()
+
+			valid, err := r.validateRepo(orm.Clone(req.Context()), repo.RepositoryID)
```
Comment (ywk253100) [inline src/server/registry/catalog.go:107]: We can call the tag manager directly to check whether the repository contains tags or not without listing artifacts
Further more, how about to introduce a function in the repository manager to get the repositories with tags directly? We can join the table repository and tag in database level and get the desired result

Comment (heww) [reply src/server/registry/catalog.go:107]: Agree to call the tag manager for checking. Not agree to do it in the repo manager using join between the repo table with the tag table.

Join is an expensive operation on large tables.

Comment (wy65701436) [reply src/server/registry/catalog.go]: use the tagCtl for instead.

File: src/server/registry/catalog.go [hunk 3]
```diff
@@ -68,16 +74,54 @@ func (r *repositoryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
...
+
+	for _, repo := range repoRecords {
+		r.wp.GetWorker()
+		go func(repo *models.RepoRecord) {
+			defer func() {
+				r.wp.ReleaseWorker()
+				wg.Done()
+			}()
+
+			tags, err := r.tagCtl.List(orm.Clone(req.Context()), &q.Query{
```
Comment (ywk253100) [inline src/server/registry/catalog.go:106]: Use `Count` instead

File: src/server/registry/catalog.go [hunk 4]
```diff
@@ -68,16 +74,54 @@ func (r *repositoryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)
...
+				Keywords: map[string]interface{}{
+					"repository_id": repo.RepositoryID,
+				},
+			}, nil)
+			if err != nil {
+				lib_http.SendError(w, err)
```
Comment (ywk253100) [inline src/server/registry/catalog.go:112]: Need to pass the error out of the goroutine and cancel all other goroutines

File: src/server/registry/catalog.go (hunk 3 above)
Comment (wy65701436) [reply src/server/registry/catalog.go]: updated

File: src/server/registry/catalog.go (hunk 4 above)
Comment (wy65701436) [reply src/server/registry/catalog.go]: updated

File: src/server/registry/catalog.go (hunk 1 above)
Comment (wy65701436) [reply src/server/registry/catalog.go]: defined a error chan

**Pull request closed without merging**